gcc/rust/expand/rust-proc-macro.cc

```diff
@@ -18,12 +18,32 @@
 #include "rust-proc-macro.h"
 #include "rust-lex.h"
 #include "rust-token-converter.h"
+#include "rust-attributes.h"
 #ifndef _WIN32
 #include <dlfcn.h>
 #endif
 
 namespace Rust {
 
+BangProcMacro::BangProcMacro (ProcMacro::Bang macro)
+  : name (macro.name),
+    node_id (Analysis::Mappings::get ()->get_next_node_id ()),
+    macro (macro.macro)
+{}
+
+AttributeProcMacro::AttributeProcMacro (ProcMacro::Attribute macro)
+  : name (macro.name),
+    node_id (Analysis::Mappings::get ()->get_next_node_id ()),
+    macro (macro.macro)
+{}
+
+CustomDeriveProcMacro::CustomDeriveProcMacro (ProcMacro::CustomDerive macro)
+  : trait_name (macro.trait_name),
+    attributes (macro.attributes, macro.attributes + macro.attr_size),
```

Comment:
what does this addition do?

Reply:
no, that's okay - I get why you added these classes. I meant, what does `macro.attributes, macro.attributes + macro.attr_size` do when creating the vector? :D

Reply:
Oh, that's for creating a vector from a pointer array, it copies every element into the vector.

Constructor (5) from the [cpp reference documentation](https://en.cppreference.com/w/cpp/container/vector/vector)

```
template< class InputIt >
vector( InputIt first, InputIt last,
        const Allocator& alloc = Allocator() );
```